patches/net/minecraft/client/multiplayer/ClientCommonPacketListenerImpl.java.patch

```diff
@@ -15,3 +21,54 @@
  
      protected abstract RegistryAccess.Frozen registryAccess();
  
+@@ -189,13 +_,24 @@
+     }
+ 
+     public void send(Packet<?> p_295097_) {
++        if (!net.neoforged.neoforge.network.registration.NetworkRegistry.getInstance().canSendPacket(p_295097_, this)) {
++            return;
++        }
++        
+         this.connection.send(p_295097_);
+     }
+ 
++    public void send(CustomPacketPayload payload) {
++        this.send(new net.minecraft.network.protocol.common.ServerboundCustomPayloadPacket(payload));
++    }
++
+     @Override
+     public void onDisconnect(Component p_295485_) {
+         this.telemetryManager.onDisconnect();
+         this.minecraft.disconnect(this.createDisconnectScreen(p_295485_));
++        if (!this.connection.isMemoryConnection()) {
++            net.neoforged.neoforge.registries.RegistryManager.revertToFrozen();
++        }
+         LOGGER.warn("Client disconnected with reason: {}", p_295485_.getString());
+     }
+ 
+@@ -315,5 +_,25 @@
+         @OnlyIn(Dist.CLIENT)
+         static record PendingRequest(UUID id, URL url, String hash) {
+         }
++    }
++    
++    @Override
++    public Connection getConnection() {
++        return connection;
++    }
++    
++    @Override
++    public net.minecraft.util.thread.ReentrantBlockableEventLoop<?> getMainThreadEventLoop() {
++        return minecraft;
++    }
++    
++    @Override
++    public ClientCommonPacketListener self() {
```

Comment:
Why is this done here rather than how we do other extension interfaces of having the `self` method be private and just cast to the expected type?

Reply:
Because the self with cast in interfaces causes a massive overhead because the compiler injects multiple instance of checks which can not be inlined or improved.